@nx/js:tsc build fails on Windows with Yarn PnP after 19.6.1

### Current Behavior

Typescript compilation fails when migrating from nx 19.6.0 to 19.6.1 on Windows with Yarn PnP
```
 NX   Running target build for project mylib

   ×  nx run mylib:build
      Compiling TypeScript files for project "mylib"...
      /git_clones/test-nx-pnp-windows/packages/mylib/src/index.ts:1:1 - error TS2354: This syntax requires an imported helper but module 'tslib' cannot be found.

      1 export * from './lib/mylib';
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      /git_clones/test-nx-pnp-windows/packages/mylib/src/lib/mylib.ts:1:25 - error TS2307: Cannot find module 'semver' or its corresponding type declarations.

      1 import * as semver from 'semver';
```

### Expected Behavior

The build should be successful as on 19.6.0

### GitHub Repo

https://github.com/fpaul-1A/test-nx-pnp-windows/tree/19.6.1

### Steps to Reproduce

1. git clone https://github.com/fpaul-1A/test-nx-pnp-windows/tree/19.6.1
2. yarn install
3. yarn build


### Nx Report

```shell
NX   Report complete - copy this into the issue template

Node           : 20.14.0
OS             : win32-x64
Native Target  : x86_64-windows
yarn           : 4.4.1

nx          : 19.6.1
@nx/js      : 19.6.1
typescript  : 5.5.4
```


### Failure Logs

_No response_

### Package Manager Version

_No response_

### Operating System

- [ ] macOS
- [ ] Linux
- [X] Windows
- [ ] Other (Please specify)

### Additional Information

The issue doesn't appear with `nodeLinker: node-modules`.
Also, it seems to be somehow related to this modification https://github.com/nrwl/nx/pull/27300, the build succeeds when reverting the change on `tsc.impl.ts`

--- a/e2e/eslint/src/linter.test.ts
+++ b/e2e/eslint/src/linter.test.ts
@@ -506,7 +506,9 @@ describe('Linter', () => {
       it('should report dependency check issues', () => {
         const rootPackageJson = readJson('package.json');
         const nxVersion = rootPackageJson.devDependencies.nx;
-        const tslibVersion = rootPackageJson.dependencies['tslib'];
+        const tslibVersion =
+          rootPackageJson.dependencies['tslib'] ||
+          rootPackageJson.devDependencies['tslib'];
 
         let out = runCLI(`lint ${mylib}`, {
           silenceError: true,
@@ -547,20 +549,18 @@ describe('Linter', () => {
           `Successfully ran target lint for project ${mylib}`
         );
         const packageJson = readJson(`libs/${mylib}/package.json`);
-        expect(packageJson).toMatchInlineSnapshot(`
-          {
-            "dependencies": {
-              "@nx/devkit": "${nxVersion}",
-              "tslib": "${tslibVersion}",
-            },
-            "main": "./src/index.js",
-            "name": "@proj/${mylib}",
-            "private": true,
-            "type": "commonjs",
-            "typings": "./src/index.d.ts",
-            "version": "0.0.1",
-          }
-        `);
+        expect(packageJson).toMatchObject({
+          dependencies: {
+            '@nx/devkit': nxVersion,
+            tslib: tslibVersion,
+          },
+          main: './src/index.js',
+          name: `@proj/${mylib}`,
+          private: true,
+          type: 'commonjs',
+          typings: './src/index.d.ts',
+          version: '0.0.1',
+        });
 
         // intentionally set the invalid version
         updateJson(`libs/${mylib}/package.json`, (json) => {

--- a/packages/js/src/generators/init/init.spec.ts
+++ b/packages/js/src/generators/init/init.spec.ts
@@ -139,4 +139,31 @@ describe('js init generator', () => {
     expect(tree.exists('.prettierignore')).toBeFalsy();
     expect(tree.exists('.prettierrc')).toBeFalsy();
   });
+
+  it.each`
+    fileName                | importHelpers | shouldAdd
+    ${'tsconfig.json'}      | ${true}       | ${true}
+    ${'tsconfig.base.json'} | ${true}       | ${true}
+    ${'tsconfig.json'}      | ${false}      | ${false}
+    ${'tsconfig.base.json'} | ${false}      | ${false}
+    ${null}                 | ${false}      | ${false}
+  `(
+    'should add tslib if importHelpers is true in base tsconfig',
+    async ({ fileName, importHelpers, shouldAdd }) => {
+      if (fileName) {
+        writeJson(tree, fileName, {
+          compilerOptions: {
+            importHelpers,
+          },
+        });
+      }
+
+      await init(tree, {
+        addTsConfigBase: false,
+      });
+
+      const packageJson = readJson(tree, 'package.json');
+      expect(!!packageJson.devDependencies?.['tslib']).toBe(shouldAdd);
+    }
+  );
 });

--- a/packages/js/src/generators/init/init.ts
+++ b/packages/js/src/generators/init/init.ts
@@ -21,6 +21,7 @@ import {
   swcCoreVersion,
   swcHelpersVersion,
   swcNodeVersion,
+  tsLibVersion,
   typescriptVersion,
 } from '../../utils/versions';
 import { InitSchema } from './schema';
@@ -108,6 +109,16 @@ export async function initGeneratorInternal(
       skipPackageJson: schema.skipPackageJson,
     });
     tasks.push(prettierTask);
+  }
+
+  const rootTsConfigFileName = getRootTsConfigFileName(tree);
+  // If the root tsconfig file uses `importHelpers` then we must install tslib
+  // in order to run tsc for build and typecheck.
+  if (rootTsConfigFileName) {
+    const rootTsConfig = readJson(tree, rootTsConfigFileName);
+    if (rootTsConfig.compilerOptions?.importHelpers) {
+      devDependencies['tslib'] = tsLibVersion;
+    }
   }
 
   const installTask = !schema.skipPackageJson

--- a/packages/js/src/generators/library/library.ts
+++ b/packages/js/src/generators/library/library.ts
@@ -170,10 +170,6 @@ export async function libraryGeneratorInternal(
     ]);
   }
 
-  if (options.bundler !== 'none') {
-    addBundlerDependencies(tree, options);
-  }
-
   if (!options.skipFormat) {
     await formatFiles(tree);
   }
@@ -409,23 +405,6 @@ export async function addLint(
     );
   }
   return task;
-}
-
-function addBundlerDependencies(tree: Tree, options: NormalizedSchema) {
-  updateJson(tree, `${options.projectRoot}/package.json`, (json) => {
-    if (options.bundler === 'tsc') {
-      json.dependencies = {
-        ...json.dependencies,
-        tslib: tsLibVersion,
-      };
-    } else if (options.bundler === 'swc') {
-      json.dependencies = {
-        ...json.dependencies,
-        '@swc/helpers': swcHelpersVersion,
-      };
-    }
-    return json;
-  });
 }
 
 function updateTsConfig(tree: Tree, options: NormalizedSchema) {

